fix OOM issue when update weight from trainer.

MayDomine opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

I encountered an out-of-memory (OOM) issue while running the grpo_gsm8k example in AReal. The error message said: "Failed to update parameter online: OOM ...". After investigating, I found that the model weight update process in sglang—specifically in `model_runner.py->ModelRunner-> update_weights_from_distributed`—may be causing the OOM.

The current implementation creates a full copy of the model weights and uses an NCCL broadcast call to load weights from the trainer. This approach significantly increases memory usage, potentially leading to OOM errors.


https://github.com/sgl-project/sglang/blob/a13dd1e492a303bba4f8ceb2e6d476797ffb052a/python/sglang/srt/model_executor/model_runner.py#L1047-L1074

To address this, I propose introducing a fake tensor with a _fake attribute to indicate when a tensor is not yet loaded. The weight loader should then load the actual weights only when the fake tensor is encountered. This change would reduce memory overhead during weight updates.
 
## Modifications

I have drafted a preliminary implementation for the default weight loader but would appreciate community assistance to finalize this pull request.

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

@zhuzilin @zhaochenyang20 CC.

MayDomine committed:
fix OOM issue when update weight from trainer.

zhaochenyang20 commented:
Great job!

zhaochenyang20 commented on python/sglang/srt/model_loader/weight_utils.py:
```diff
@@ -637,10 +637,23 @@ def convert_pyslice_to_tensor(x: Any) -> torch.Tensor:
         x = x[:]
     return x
 
+def bcast_weight(weight):
+    assert weight._original_shape and weight._group
+    real_weight = torch.empty(weight._original_shape, dtype=weight.dtype, device=weight.device)
+
+    torch.distributed.broadcast(
+        real_weight,
+        src=0,
+        group=weight._group,
+        async_op=False,
+    )
+    return real_weight
```
gemini review is great. And, please use `broadcast_weight` as the function name.

> ![medium](https://camo.githubusercontent.com/a33c7ba47779a2aabda978d90fc2665dd949d88da87f5ff554e65308de059f02/68747470733a2f2f7777772e677374617469632e636f6d2f636f64657265766965776167656e742f6d656469756d2d7072696f726974792e737667)
> 
> For better code clarity, readability, and maintainability, it would be beneficial to add type hints and a docstring to the new `bcast_weight` function. This will make it easier for future developers to understand its purpose, arguments, and what it returns, especially since it operates on tensors with custom attributes. Also, using `hasattr` for the assertion would be more explicit about checking for attribute existence.
> 
> ```python
> def bcast_weight(weight: torch.Tensor) -> torch.Tensor:
>     """Create a real tensor and broadcast weights from the source.
> 
>     This function is used to handle "fake" tensors which are placeholders
>     to avoid large memory allocation before the actual weight loading.
> 
>     Args:
>         weight: The fake tensor containing metadata (_original_shape, _group).
> 
>     Returns:
>         The real tensor with weights broadcasted from the source.
>     """
>     assert hasattr(weight, "_original_shape") and hasattr(weight, "_group")
>     real_weight = torch.empty(weight._original_shape, dtype=weight.dtype, device=weight.device)
> 
>     torch.distributed.broadcast(
>         real_weight,
>         src=0,
>         group=weight._group,
>         async_op=False,
>     )
>     return real_weight
> ```

gemini review is great. And, please use `broadcast_weight` as the function name.

zhaochenyang20 commented:
fix the lint please.

MayDomine committed:
use a universal hook function to handle all weight loader

Meanwhile this commit use broadcast_weight as function name and fixed
the lint.

zhaochenyang20 commented:
@MayDomine fix the lint? 🤔

MayDomine committed:
fix the lint

MayDomine committed:
fix the statement

zhaochenyang20 reviewed (approved):
LGTM, but pass the CI and fix lint, rebase it

MayDomine committed:
fix import

MayDomine committed:
fix import

MayDomine commented:
> "I have two questions:
> 
> Regarding the OOM: Is it caused by updating too many weights at once? Can we group/chunk the updates (similar to [slime](https://github.com/THUDM/slime/blob/805ffee542afb17a048f56d4149f78da71646219/slime/backends/megatron_utils/update_weight/update_weight_from_distributed.py#L137)) to avoid OOM?
> 
> Regarding performance: If I change the logic to broadcast(async=False) > load.weights > broadcast(async=False) > load.weights, will it be slower compared to the previous approach of allbroadcast(async=False) > all load.weights?"



> "I have two questions:
> 
> Regarding the OOM: Is it caused by updating too many weights at once? Can we group/chunk the updates (similar to [slime](https://github.com/THUDM/slime/blob/805ffee542afb17a048f56d4149f78da71646219/slime/backends/megatron_utils/update_weight/update_weight_from_distributed.py#L137)) to avoid OOM?
> 
> Regarding performance: If I change the logic to broadcast(async=False) > load.weights > broadcast(async=False) > load.weights, will it be slower compared to the previous approach of allbroadcast(async=False) > all load.weights?"

lilei199908 commented:
> > "I have two questions:
> 
> > 
> 
> > Regarding the OOM: Is it caused by updating too many weights at once? Can we group/chunk the updates (similar to [slime](https://github.com/THUDM/slime/blob/805ffee542afb17a048f56d4149f78da71646219/slime/backends/megatron_utils/update_weight/update_weight_from_distributed.py#L137)) to avoid OOM?
> 
> > 
> 
> > Regarding performance: If I change the logic to broadcast(async=False) > load.weights > broadcast(async=False) > load.weights, will it be slower compared to the previous approach of allbroadcast(async=False) > all load.weights?"
> 
> 
> 
> 
> 
> 
> 
> For Q1: While using Slime, I also encountered this OOM issue and confirmed that this PR resolves it. I suspect that Slime's current bucketing/chunking logic may have a problem with memory not being released in a timely manner, which requires further testing. Additionally, the bucket size might also be a contributing factor. Currently, with the current settings, Qwen3-8B on 8 H100 GPUs, Slime's existing bucketing logic has not resolved the issue.
> 
> 
> 
> For Q2: Indeed, it would be slower in this case. I think we might need a switch for better control between memory saving and efficiency.
> 
> 

Get it，thank you for your reply！
